R/restart.R

```diff
@@ -59,10 +62,22 @@ renv_restart_request_rstudio <- function(project, reason, ...) {
     restart <- getOption("renv.restart.function", default = function() {
       tools$.rs.api.executeCommand("restartR", quiet = TRUE)
     })
-    return(invisible(restart()))
+    return(renv_restart_invoke(restart))
   }
 
   # otherwise, explicitly open the new project
-  invisible(tools$.rs.api.openProject(project, newSession = FALSE))
+  renv_restart_invoke(function() {
+    invisible(tools$.rs.api.openProject(project, newSession = FALSE))
+  })
+
+}
+
+renv_restart_invoke <- function(callback) {
+
+  # avoid multiple attempts to restart in a single call, just in case
+  if (!the$restarting) {
+    the$restarting <- TRUE
+    callback()
+  }
```

Comment:
```suggestion
  }
  
  the$restarting <- FALSE
```

Just in case it fails?

Reply:
The problem here is that the restart will happen "eventually", so we don't know directly in R code if the restart was (or will be) successful at that point.

Reply:
Yeah, I wondered about that. I just worry about getting stuck in some bad state if the restart fails.